libraries/chain/block_log.cpp

```diff
@@ -615,6 +621,63 @@ namespace eosio { namespace chain {
             FC_LOG_AND_RETHROW()
          }
 
+         std::vector<char> read_serialized_block_by_num(uint32_t block_num) final {
+            try {
+               uint64_t pos = get_block_pos(block_num);
+
+               if (pos == block_log::npos || !head) {
+                  // Rare case. No need a special code for it.
+                  // Just fall back to the regular `retry_read_block_by_num` and then serialize.
+                  auto ret = retry_read_block_by_num(block_num);
+                  if (ret) {
+                     return fc::raw::pack(*ret);
+                  } else {
+                     return {};
+                  }
+               }
+
+               uint32_t head_block_num = block_header::num_from_id(head->id);
+               EOS_ASSERT(block_num <= head_block_num, block_log_exception,
+                          "Current block_num ${n} was greater than head_block_num ${h}",
+                          ("n", block_num)("h", head_block_num));
+
+               uint64_t block_size     = 0;
+               constexpr uint32_t block_pos_size = sizeof(uint64_t); // size of block position field in the block log file
+
+               if (block_num < head_block_num) {
+                  // current block is not the last block in the log file.
+                  uint64_t next_block_pos = get_block_pos(block_num + 1);
+
+                  EOS_ASSERT(next_block_pos != block_log::npos, block_log_exception,
+                             "no position found for block ${n}", ("n", block_num + 1));
+                  EOS_ASSERT(next_block_pos > pos + block_pos_size, block_log_exception,
+                             "next block position ${np} should be greater than current block position ${p} plus block position field size ${bps}",
+                             ("np", next_block_pos)("p", pos)("bps", block_pos_size));
+
+                  block_size = next_block_pos - pos - block_pos_size;
+               } else {
+                  // current block is the last block in the file.
+                  assert (block_num == head_block_num);
+
+                  auto file_size = std::filesystem::file_size(block_file.get_file_path());
```

Comment:
why not use `size()`

Reply:
`basic_block_log` does not have `size()` to return the size of its block log file.